File: main/SectionType.cpp
#include "SectionType.h"

enum class SectionType::SectionEnum {
    HEAD, CENTER_TORSO, RIGHT_TORSO, LEFT_TORSO, REAR_CENTER_TORSO, REAR_RIGHT_TORSO, REAR_LEFT_TORSO, RIGHT_ARM, LEFT_ARM, RIGHT_LEG,
    LEFT_LEG, FRONT_RIGHT_LEG, FRONT_LEFT_LEG, REAR_RIGHT_LEG, REAR_LEFT_LEG, SPECIAL_1, SPECIAL_2
};

SectionType::SectionType(const SectionType::SectionEnum & sectionType, const std::string & fullName, const std::string & abbreviation)
    :longName(fullName), shortName(abbreviation), enumId(sectionType) {
}

const SectionType SectionType::HEAD(SectionType::SectionEnum::HEAD, "Head", "HD");
const SectionType SectionType::CENTER_TORSO(SectionType::SectionEnum::CENTER_TORSO, "Head", "CT");
const SectionType SectionType::RIGHT_TORSO(SectionType::SectionEnum::RIGHT_TORSO, "Head", "RT");
const SectionType SectionType::LEFT_TORSO(SectionType::SectionEnum::LEFT_TORSO, "Head", "LT");
const SectionType SectionType::REAR_CENTER_TORSO(SectionType::SectionEnum::REAR_CENTER_TORSO, "Head", "RCT");
const SectionType SectionType::REAR_RIGHT_TORSO(SectionType::SectionEnum::REAR_RIGHT_TORSO, "Head", "RRT");
const SectionType SectionType::REAR_LEFT_TORSO(SectionType::SectionEnum::REAR_LEFT_TORSO, "Head", "RLT");
const SectionType SectionType::RIGHT_ARM(SectionType::SectionEnum::RIGHT_ARM, "Head", "RA");
const SectionType SectionType::LEFT_ARM(SectionType::SectionEnum::LEFT_ARM, "Head", "LA");
const SectionType SectionType::RIGHT_LEG(SectionType::SectionEnum::RIGHT_LEG, "Head", "RL");
const SectionType SectionType::LEFT_LEG(SectionType::SectionEnum::LEFT_LEG, "Head", "LL");
const SectionType SectionType::FRONT_RIGHT_LEG(SectionType::SectionEnum::FRONT_RIGHT_LEG, "Head", "FRL");
const SectionType SectionType::FRONT_LEFT_LEG(SectionType::SectionEnum::FRONT_LEFT_LEG, "Head", "FLL");
const SectionType SectionType::REAR_RIGHT_LEG(SectionType::SectionEnum::REAR_RIGHT_LEG, "Head", "RRL");
const SectionType SectionType::REAR_LEFT_LEG(SectionType::SectionEnum::REAR_LEFT_LEG, "Head", "RLL");
const SectionType SectionType::SPECIAL_1(SectionType::SectionEnum::SPECIAL_1, "Head", "S1");
const SectionType SectionType::SPECIAL_2(SectionType::SectionEnum::SPECIAL_2, "Head", "S2");

File: src/main/java/com/github/majora_incarnate/mwo_mechbay/entities/SectionType.java
package com.github.majora_incarnate.mwo_mechbay.entities;

/**
 * @author Trevin
 */
public enum SectionType {
    UNKNOWN("NA", "Unknown", false, -1),
    HEAD("HD", "Head", false, 0),
    LEFT_ARM("LA", "Left Arm", false, 1),
    RIGHT_ARM("RA", "Right Arm", false, 2),
    LEFT_TORSO("LT", "Left Torso", true, 3),
    RIGHT_TORSO("RT", "Right Torso", true, 4),
    CENTER_TORSO("CT", "Center Torso", true, 5),
    LEFT_LEG("LL", "Left Leg", false, 6),
    RIGHT_LEG("RL", "Right Leg", false, 7),
    SPECIAL_ONE("S1", "Special 1, false", 8),
    SPECIAL_TWO("S2", "Special 2, false", 9),
    FRONT_LEFT_LEG("FLL", "Front Left Leg", false, 1),
    FRONT_RIGHT_LEG("FRL", "Front Right Leg", false, 2),
    REAR_LEFT_LEG("RLL", "Rear Left Leg", false, 6),
    REAR_RIGHT_LEG("RRL", "Rear Right Leg", false, 7);

    public final String shortName;
    public final String longName;
    public final boolean hasRear;
    public final int index;

    SectionType(String shortName, String longName, boolean hasRear, int index) {
        this.shortName = shortName;
        this.longName = longName;
        this.hasRear = hasRear;
        this.index = index;
    }

    public int count() {
        return SectionType.values().length;
    }

    @Override
    public String toString() {
        return longName;
    }
    
    public static int numberOfSectionTypes() {
        return RIGHT_LEG.index + 1;
    }

    public static SectionType getEnum(String value) {
        switch(value) {
            case "HD":
                return HEAD;
            case "LA":
                return LEFT_ARM;
            case "RA":
                return RIGHT_ARM;
            case "LT":
                return LEFT_TORSO;
            case "RT":
                return RIGHT_TORSO;
            case "CT":
                return CENTER_TORSO;
            case "LL":
                return LEFT_LEG;
            case "RL":
                return RIGHT_LEG;
            case "FLL":
                return FRONT_LEFT_LEG;
            case "FRL":
                return FRONT_RIGHT_LEG;
            case "RLL":
                return REAR_LEFT_LEG;
            case "RRL":
                return REAR_RIGHT_LEG;
            default:
                return UNKNOWN;
        }
    }
}
